// @stricahq/typhonjs/dist/internal-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutputItemType = exports.TransactionBodyItemType = exports.RedeemerTag = void 0;
var RedeemerTag;
(function (RedeemerTag) {
    RedeemerTag[RedeemerTag["SPEND"] = 0] = "SPEND";
    RedeemerTag[RedeemerTag["MINT"] = 1] = "MINT";
    RedeemerTag[RedeemerTag["CERT"] = 2] = "CERT";
    RedeemerTag[RedeemerTag["REWARD"] = 3] = "REWARD";
})(RedeemerTag || (exports.RedeemerTag = RedeemerTag = {}));
var TransactionBodyItemType;
(function (TransactionBodyItemType) {
    TransactionBodyItemType[TransactionBodyItemType["INPUTS"] = 0] = "INPUTS";
    TransactionBodyItemType[TransactionBodyItemType["OUTPUTS"] = 1] = "OUTPUTS";
    TransactionBodyItemType[TransactionBodyItemType["FEE"] = 2] = "FEE";
    TransactionBodyItemType[TransactionBodyItemType["TTL"] = 3] = "TTL";
    TransactionBodyItemType[TransactionBodyItemType["CERTIFICATES"] = 4] = "CERTIFICATES";
    TransactionBodyItemType[TransactionBodyItemType["WITHDRAWALS"] = 5] = "WITHDRAWALS";
    TransactionBodyItemType[TransactionBodyItemType["AUXILIARY_DATA_HASH"] = 7] = "AUXILIARY_DATA_HASH";
    TransactionBodyItemType[TransactionBodyItemType["VALIDITY_INTERVAL_START"] = 8] = "VALIDITY_INTERVAL_START";
    TransactionBodyItemType[TransactionBodyItemType["MINT"] = 9] = "MINT";
    TransactionBodyItemType[TransactionBodyItemType["SCRIPT_DATA_HASH"] = 11] = "SCRIPT_DATA_HASH";
    TransactionBodyItemType[TransactionBodyItemType["COLLATERAL_INPUTS"] = 13] = "COLLATERAL_INPUTS";
    TransactionBodyItemType[TransactionBodyItemType["REQUIRED_SIGNERS"] = 14] = "REQUIRED_SIGNERS";
    TransactionBodyItemType[TransactionBodyItemType["NETWORK_ID"] = 15] = "NETWORK_ID";
    TransactionBodyItemType[TransactionBodyItemType["COLLATERAL_OUTPUT"] = 16] = "COLLATERAL_OUTPUT";
    TransactionBodyItemType[TransactionBodyItemType["TOTAL_COLLATERAL"] = 17] = "TOTAL_COLLATERAL";
    TransactionBodyItemType[TransactionBodyItemType["REFERENCE_INPUTS"] = 18] = "REFERENCE_INPUTS";
})(TransactionBodyItemType || (exports.TransactionBodyItemType = TransactionBodyItemType = {}));
var OutputItemType;
(function (OutputItemType) {
    OutputItemType[OutputItemType["ADDRESS"] = 0] = "ADDRESS";
    OutputItemType[OutputItemType["VALUE"] = 1] = "VALUE";
    OutputItemType[OutputItemType["DATUM_OPTION"] = 2] = "DATUM_OPTION";
    OutputItemType[OutputItemType["SCRIPT_REF"] = 3] = "SCRIPT_REF";
})(OutputItemType || (exports.OutputItemType = OutputItemType = {}));
// NativeScript types end
