// @stricahq/typhonjs/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkId = exports.WitnessType = exports.CertificateType = exports.PlutusScriptType = exports.HashType = void 0;
var HashType;
(function (HashType) {
    HashType[HashType["ADDRESS"] = 0] = "ADDRESS";
    HashType[HashType["SCRIPT"] = 1] = "SCRIPT";
})(HashType || (exports.HashType = HashType = {}));
var PlutusScriptType;
(function (PlutusScriptType) {
    PlutusScriptType["PlutusScriptV1"] = "PlutusScriptV1";
    PlutusScriptType["PlutusScriptV2"] = "PlutusScriptV2";
})(PlutusScriptType || (exports.PlutusScriptType = PlutusScriptType = {}));
var CertificateType;
(function (CertificateType) {
    CertificateType[CertificateType["STAKE_REGISTRATION"] = 0] = "STAKE_REGISTRATION";
    CertificateType[CertificateType["STAKE_DE_REGISTRATION"] = 1] = "STAKE_DE_REGISTRATION";
    CertificateType[CertificateType["STAKE_DELEGATION"] = 2] = "STAKE_DELEGATION";
})(CertificateType || (exports.CertificateType = CertificateType = {}));
var WitnessType;
(function (WitnessType) {
    WitnessType[WitnessType["V_KEY_WITNESS"] = 0] = "V_KEY_WITNESS";
    WitnessType[WitnessType["NATIVE_SCRIPT"] = 1] = "NATIVE_SCRIPT";
    WitnessType[WitnessType["PLUTUS_SCRIPT_V1"] = 3] = "PLUTUS_SCRIPT_V1";
    WitnessType[WitnessType["PLUTUS_DATA"] = 4] = "PLUTUS_DATA";
    WitnessType[WitnessType["REDEEMER"] = 5] = "REDEEMER";
    WitnessType[WitnessType["PLUTUS_SCRIPT_V2"] = 6] = "PLUTUS_SCRIPT_V2";
})(WitnessType || (exports.WitnessType = WitnessType = {}));
var NetworkId;
(function (NetworkId) {
    NetworkId[NetworkId["MAINNET"] = 1] = "MAINNET";
    NetworkId[NetworkId["TESTNET"] = 0] = "TESTNET";
})(NetworkId || (exports.NetworkId = NetworkId = {}));
