// entityos/entityos.js
var _ = require('lodash')
var moment = require('moment');

module.exports = 
{
	VERSION: '2.2.11',

	data: {session: undefined},
	controller: {},

	init: function (callBack, settings)
	{
		if (_.isPlainObject(settings))
		{
			module.exports.data._settings = settings;
					
			if (!_.isUndefined(settings.entityos))
			{
				settings = settings.entityos;
			}

			module.exports.data.settings = settings;
			module.exports.logon(callBack, settings);
		}	
		else
		{	
			var siteSuffix = '';

			if (_.isString(settings))
			{
				siteSuffix = '-' + settings;
			}
			else
			{
				var _event = module.exports.data._event;

				if (_.has(_event, 'site'))
				{
					siteSuffix = '-' + _event.site;
				}

				if (_.has(_event, 'context'))
				{
					siteSuffix = '-' + _event.context;
				}

				if (_.has(_event, 'space'))
				{
					siteSuffix = '-' + _event.space;
				}

				if (_.has(_event, 'settings'))
				{
					siteSuffix = '-' + _event.settings;
				}

				if (_.has(_event, 'body.context'))
				{
					siteSuffix = '-' +_event.body.context;
				}

				if (_.has(_event, 'body.data._context'))
				{
					siteSuffix = '-' +_event.body.data._context;
				}
			}

			var fs = require('fs');

			fs.readFile('settings' + siteSuffix + '.json', function (err, buffer)
			{
				if (!err)
				{	
					var _settings = buffer.toString();
					settings = JSON.parse(_settings);

					if (_.has(settings, 'rules'))
					{
						_.each(settings.rules, function (rule)
						{
							if (_.has(rule, 'set'))
							{
								if (rule.value == undefined)
								{ 
									if (rule.env != undefined)
									{
										rule.value = process.env[rule.env];
										
										if (_.isUndefined(rule.value))
										{
											rule.value = rule.default;
										}
									}
								}

								_.set(settings, rule.set, rule.value)
							}
						});
					}

					module.exports.data._settings = settings;

					if (!_.isUndefined(settings))
					{
						if (!_.isUndefined(settings.entityos))
						{
							settings = settings.entityos;
						}

						module.exports.data.settings = settings;
					}

					var logon = (settings.logon != undefined);
					module.exports._util.testing.data(logon, 'entityos-init|entityos-cloud-logon');

					if (logon)
					{
						module.exports.logon(callBack, settings)
					}
					else
					{
						if (_.isFunction(callBack))
						{
							callBack({settings: settings})
						}
						else
						{
							module.exports._util.controller.invoke(callBack, {settings: settings}, settings);
						}
					}
				}
				else
				{
					console.log('ERROR! No settings' + siteSuffix + '.json file.')
				}
			});
		}
	},

	logon: function (callBack, settings, param)
	{
		var https = require('https');
		var authenticationLevel = module.exports.data.settings.authenticationLevel;
		if (_.isUndefined(authenticationLevel)) {authenticationLevel = 1};

		settings = _.assign(module.exports.data.settings, settings);
		
		var logon = settings.logon;
		var password = settings.password;
		var code = module.exports._util.param.get(param, 'code').value;

		var data = 
		{
			method: 'LOGON',
			logon: logon,
			localtime: moment().format('D MMM YYYY HH:mm:ss')
		}

		var requestData = 'logon=' + logon +
								'&localtime=' + moment().format('D MMM YYYY HH:mm:ss');

		if (authenticationLevel == 1)
		{
			requestData += '&passwordhash=' + module.exports._util.hash(logon + password);
		}
		else if (authenticationLevel == 2)
		{
			requestData += '&passwordhash=' + module.exports._util.hash(logon + password + module.exports.data.session.logonkey)
		}
		else if (authenticationLevel == 3)
		{
			requestData += '&passwordhash=' + module.exports._util.hash(logon + password + module.exports.data.session.logonkey + code)
		}

		module.exports._util.testing.data(requestData, 'entityos.cloud.logon.request');
		
		var options =
		{
			hostname: settings.hostname,
			port: 443,
			path: '/rpc/logon/?method=LOGON',
			method: 'POST',
			headers:
			{
				'Content-Type': 'application/x-www-form-urlencoded',
				'Content-Length': requestData.length
			}
		};

		module.exports._util.testing.data(options, 'entityos.cloud.send.request.options');

		var req = https.request(options, function(res)
		{
			res.setEncoding('utf8');

			var data = '';
			
			res.on('data', function(chunk)
			{
			  	data += chunk;
			});
			
			res.on('end', function ()
			{	
				module.exports._util.testing.data(data, 'entityos.cloud.send.response.end');
				var _data = JSON.parse(data);
		    	module.exports.data.session = _data;
		    	module.exports.data._session = _data;
		    	
		    	if (_.isFunction(callBack))
		    	{
		    		callBack({data: _data, settings: settings})
		    	}
		    	else
		    	{
		    		module.exports._util.controller.invoke(callBack, _data, settings);
		    	}
			});
		});

		req.on('error', function(error)
		{
			module.exports._util.testing.data(error.message, 'entityos.cloud.logon.response.error');
		 
		  	if (_.isFunction(callBack))
		  	{
		  		callBack({error: error})
		  	}
	  		else
	    	{
	    		module.exports._util.controller.invoke(callBack, options, {error: error})
	    	}
		});

		module.exports._util.testing.data(requestData, 'entityos.cloud.send.request.data');

		req.write(requestData);
		req.end()
	},

	logoff: function (param, response)
	{
		var uri = module.exports._util.param.get(param, 'uri', {"default": '/'}).value;
		var refresh = module.exports._util.param.get(param, 'refresh', {"default": true}).value;

		if (_.isUndefined(response))
		{
			module.exports.send(
			{
				url: '/rpc/core/?method=CORE_LOGOFF'
			},
			'',
			module.exports._util.logoff);
		}
		else
		{
			module.exports._util.testing.log('Logged off', 'entityos.cloud._util.logoff');
			module.exports.data.session = undefined;
		}	
	},		

	send: function (options, data, callBack, callBackParam)
	{
		//Send to entityos.cloud

		var https = require('https');
		var querystring = require('querystring');
		var settings = module.exports.data.settings;
		var session = module.exports.data.session;
		var requestData;
        var headers = {};

        headers['auth-sid'] = session.sid;
        headers['auth-logonkey'] = session.logonkey;
        
        if (options == undefined) {options = {}};

		if (_.isUndefined(data))
		{
			requestData = options.data
		}
		else
		{
			requestData = data
		}

        if (options.contentType == undefined)
        {
            options.contentType = 'application/x-www-form-urlencoded'
        }

		if (_.isPlainObject(requestData))
		{
            if (options.contentType == 'application/json')
            {
                requestData = JSON.stringify(requestData);
            }
            else
            {
                var _requestData = []

                for (key in requestData)
                {
                    _requestData.push(key + '=' + requestData[key]);
                }
                
                requestData = _.join(_requestData, '&');                
            }

            headers['Content-Length'] = requestData.length;
		}
		else if (requestData != undefined)
		{    
            headers['Content-Length'] = requestData.length;
		}
		else 
		{   
			requestData = ''; 
            headers['Content-Length'] = 0;
		}

        headers['Content-Type'] = options.contentType;

		if (_.isUndefined(callBack))
		{
			callBack = options.callBack
		}

		if (_.isUndefined(callBackParam))
		{
			callBackParam = options.callBackParam
		}					
				
		if (options.type == undefined) {options.type = 'POST'}
				
		var requestOptions =
		{
			hostname: settings.hostname,
			port: 443,
			path: options.url,
			method: options.type,
			headers: headers
		};

		module.exports._util.testing.data(requestOptions, 'entityos.cloud.send.request.options');

		var req = https.request(requestOptions, function(res)
		{
			res.setEncoding('utf8');

			var data = '';
			
			res.on('data', function(chunk)
			{
			  	data += chunk;
			});
			
			res.on('end', function ()
			{	
				module.exports._util.testing.data(data, 'entityos.cloud.send.response.end');

				if (_.startsWith(data, '{') && (settings.convert == 'true' || options.all == true))
				{
					data = JSON.parse(data)
				}

				if (options.all)
				{
					if (module.exports.data._send == undefined)
					{
						module.exports.data._send = {}
					}

					if (module.exports.data._send[data.moreid] == undefined)
					{
						module.exports.data._send[data.moreid] = data.data.rows
					}
					else
					{
						module.exports.data._send[data.moreid] = 
							_.concat(module.exports.data._send[data.moreid], data.data.rows)
					}
				}

				if (options.all && data.morerows == 'true')
				{
                    if (options.rows == undefined) {options.rows = 50}

					module.exports.send(
					{
						url: '/rpc/core/?method=CORE_MORE_SEARCH',
						all: true,
                        rows: options.rows,
						callBack: callBack,
						callBackParam: callBackParam
					},
					'id=' + data.moreid + '&startrow=' + _.size(module.exports.data._send[data.moreid])
					)
				}
				else
				{
					if (callBackParam != undefined)
					{
						options = callBackParam
					}

					if (options.all)
					{
						data.data.rows = module.exports.data._send[data.moreid];
					}

			    	if (_.isFunction(callBack))
			    	{
			    		callBack(options, data)
			    	}
			    	else
			    	{
			    		module.exports._util.controller.invoke(callBack, options, data);
			    	}

			    	if (options.all)
			    	{
			    		delete module.exports.data._send[data.moreid];
			    	}
			   }
			});
		});

		req.on('error', function(error)
		{
			module.exports._util.testing.message(error.message, 'entityos.cloud.send.response.error');
		  	if (_.isFunction(callBack))
		  	{
		  		callBack(options, {error: error})
		  	}
	  		else
	    	{
	    		module.exports._util.controller.invoke(callBack, options, {error: error})
	    	}
		});

		module.exports._util.testing.data(requestData, 'entityos.cloud.send.request.data');

		req.write(requestData);
		req.end()
	},

	cloud:
	{
		logon: function (param) {module.exports.logon(param)},
		logoff: function (param) {module.exports.logoff(param)},
		send: function (param, data, callBack) {module.exports.send(param, data, callBack)},
		invoke: function (param, data, callBack, callBackParam)
		{
			if (arguments.length == 1)
			{
				data = param.data;
				if (data == undefined) {data = param.fields}

				callBack = param.callback;
				callBackParam = param.callbackParam;
			}

			if (param.url == undefined && param.method != undefined)
			{
				param.url = '/rpc/' + param.method.split('_')[0] + '/?method=' + (param.method).toUpperCase();
			}

			if (param.url == undefined && param.object != undefined)
			{
				param.url = '/rpc/' + param.object.split('_')[0] + '/?method=' + (param.object).toUpperCase() + '_MANAGE';
			}

			if (_.isObject(data))
			{
				data = _.join(_.map(data, function (value, key) {return key + '=' + value}), '&')
			}

			module.exports.send(param, data, callBack, callBackParam)
		},
		create: function (param)
		{
			if (typeof arguments[0] != 'object')
			{
				param =
				{
					object: arguments[0],
					data: arguments[1],
					callback: arguments[2],
					mode: arguments[3],
					callbackParam: arguments[4]
				}
			}

			if (param.data == undefined && param.fields != undefined)
			{
				param.data = param.fields;
			}

			var endpoint = param.object.split('_')[0];
			var contentType;

			if (_.isObject(param.data))
			{
				contentType = 'application/json'
			}

			module.exports.send(
			{
				type: 'post',
				url: '/rpc/' + endpoint + '/?method=' + (param.object).toUpperCase() + '_MANAGE',
				contentType: contentType
			},
			param.data,
			param.callback,
			param.callbackParam);
		},
		update: function (param) {module.exports.cloud.create(param)},
		save: function (param) {module.exports.cloud.create(param)},

		delete: function (param)
		{
			if (typeof param.data != 'object')
			{
				param.data = {id: param.data}
			}

			param.data.remove = 1;

			if (_.isObject(param.data))
			{
				param.data = _.join(_.map(param.data, function (data, key) {return key + '=' + data}), '&')
			}

			var endpoint = param.object.split('_')[0];
			
			module.exports.send(
			{
				type: 'post',
				url: '/rpc/' + endpoint + '/?method=' + (param.object).toUpperCase() + '_MANAGE'
			},
			param.data,
			param.callback,
			param.callbackParam);
		},

		retrieve: function (param)
		{
			param.endpoint = param.object.split('_')[0];	

			if (_.has(param.data, 'criteria'))
			{
				if (_.has(param.data.criteria, 'options'))
				{
					if (_.isUndefined(param.data.criteria.options.rows))
					{
						param.data.criteria.options.rows = entityos._scope.app.options.rows;
					}
				}
				else
				{
					param.data.criteria.options = {rows: entityos._scope.app.options.rows}
				}
			}
			else
			{
				if (param.data == undefined && (_.has(param, 'fields') || _.has(param, 'summaryFields')))
				{
					param.data =
					{ 
						criteria: 
						{
							"fields": [],
							"summaryFields": [],
							"filters": [],
							"sorts": [],
							"options": {},
							"customoptions": []
						}
					}

					if (_.has(param, 'fields'))
					{ 
						if (param.fields != undefined)
						{
							if (_.isArray(param.fields))
							{
								if (_.isPlainObject(_.first(param.fields)))
								{
									param.data.criteria.fields = param.fields;
								}
								else
								{
									param.data.criteria.fields = _.map(param.fields, function (field) {return {name: field}});
								}
							}
						}
					}

					if (_.has(param, 'sorts'))
					{ 
						if (_.isArray(param.sorts))
						{
							if (_.isPlainObject(_.first(param.sorts)))
							{
								_.each(param.sorts, function (sort)
								{
									if (sort.direction == undefined)
									{
										sort.direction = 'asc'
									}

									if (sort.name == undefined)
									{
										sort.name = sort.field
									}
								});
							}
							else
							{
								param.sorts = _.map(param.sorts, function (sort) {return {name: sort, direction: 'asc'}});
							}
						}

						param.data.criteria.sorts = param.sorts
					}

					if (_.has(param, 'summaryFields')) { param.data.criteria.summaryFields = param.summaryFields }
					if (_.has(param, 'filters')) { param.data.criteria.filters = param.filters }
					if (_.has(param, 'options')) { param.data.criteria.options = param.options }
					if (_.has(param, 'customOptions')) { param.data.criteria.customOptions = param.customOptions }
					if (_.has(param, 'rows')) { param.data.criteria.options.rows = param.rows }
				}
				else
				{
					param.managed = false;
				}
			}

			if (param.data.criteria.filters != undefined)
			{
				_.each(param.data.criteria.filters, function (filter)
				{
					if (filter.value1 == undefined && filter.value != undefined)
					{
						filter.value1 = filter.value
					}

					if (filter.name == undefined && filter.field != undefined)
					{
						filter.name = filter.field
					}

					if (filter.comparison == undefined)
					{
						filter.comparison = 'EQUAL_TO'
					}
				})
			}

			param.type = 'post';
			param.url = '/rpc/' + param.endpoint + '/?method=' + (param.object).toUpperCase() + '_SEARCH';

			if (_.has(param.data, '_controller'))
			{
				param.data._controller = param.object + ':' + JSON.stringify(param.data.criteria.fields);
			}

			if (param.contentType == undefined)
			{
				param.contentType = 'application/json'
			}

			if (param.json == true)
			{
				module.exports.send(
				{
					type: param.type,
					url: param.url,
					all: param.all,
					rows: param.rows,
					contentType: param.contentType
				},
				param.data.criteria,
				param.callback,
				param.callbackParam);
			}
			else
			{
				module.exports.send(
				{
					type: param.type,
					url: param.url,
					all: param.all
				},
				'criteria=' + JSON.stringify(param.data.criteria),
				param.callback,
				param.callbackParam);
			}
		},

		search: function (param) {module.exports.cloud.retrieve(param)},
		query: function (param) {module.exports.cloud.retrieve(param)}
	},

	_util:
	{
		message: function (data, context, linebreak)
		{
			if (!_.isUndefined(context)) {console.log('[' + context + ']:')}

			if (_.isArray(data))
			{
				var _data = [];

				_.each(data, function (d)
				{
					if (_.isObject(d))
					{
						_data.push(JSON.stringify(d, null, 4))
					}
					else
					{
						_data.push(d)
					}
				})

				console.log(_.join(_data, '\n'))
			}
			else if (_.isObject(data))
			{
				console.log(JSON.stringify(data, null, 4))
			}
			else
			{
				console.log(data);
			}

			if (!_.isUndefined(linebreak))
			{
				console.log(linebreak)
			}
		},

		testing:
		{
			message: function (message, context)
			{
				if (_.has(module.exports, 'data.settings'))
				{
					var settings = module.exports.data.settings.testing;

					if (_.isObject(settings))
					{
						if (settings.status == 'true')
						{
							if (!_.isUndefined(context)) {console.log('[' + context + ']:')}
							console.log(message);

							if (!_.isUndefined(settings.break))
							{
								console.log(settings.break)
							}
						}
					}
				}
			},

			data: function (data, context)
			{
				if (_.has(module.exports, 'data.settings'))
				{
					var settings = module.exports.data.settings.testing;

					if (_.isObject(settings))
					{
						if (settings.status == 'true' && settings.showData == 'true')
						{
							if (!_.isUndefined(context)) {console.log('[' + context + '][data]:')}

							if (_.isObject(data))
							{
								console.log(JSON.stringify(data, null, 4))
							}
							else
							{
								console.log(data);
							}

							if (!_.isUndefined(settings.break))
							{
								console.log(settings.break)
							}	
						}
					}
				}
			},

			status: function (status)
			{
				if (_.has(module.exports, 'data.settings'))
				{
					var settings = module.exports.data.settings.testing;

					if (_.isObject(settings))
					{
						if (_.isUndefined(status))
						{
							status = module.exports.data.settings.testing.status;
						}
						else
						{
							module.exports.data.settings.testing.status = status;
						}
					}

					return status;
				}
			}
		},

		search:  
		{	
			init: function ()
			{
				var criteria = 
				{
					"fields": [],
					"summaryFields": [],
					"filters": [],
					"sorts": [],
					"options": {},
					"customoptions": []
				}

				return criteria
			}
		},

		hash: function (data)
		{
			var crypto = require('crypto');
			return crypto.createHash('md5').update(data).digest("hex");
		},

		doCallBack: function()
		{
			var param, callback, data;

			if (typeof arguments[0] == 'function')
			{
				callback = arguments[0]
				param = arguments[1] || {};
				data = arguments[2]
			}
			else
			{
				param = arguments[0] || {};
				callback = param.callback;
				data = arguments[1];
				delete param.callback;
			}

			if (callback)
			{
				callback(param, data)
			};
		},

		onComplete: function (param)
		{
			if (module.exports._util.param.get(param, 'onComplete').exists)
			{
				var onComplete = module.exports._util.param.get(param, 'onComplete').value;

				if (module.exports._util.param.get(param, 'onCompleteWhenCan').exists)
				{
					param.onComplete = param.onCompleteWhenCan;
					delete param.onCompleteWhenCan;
				}	
				else
				{
					delete param.onComplete;
				}

				if (typeof onComplete == 'function')
				{
					onComplete(param);
				}
				else
				{
					module.exports._util.controller.invoke(onComplete, param)
				}
			}
			else if (module.exports._util.param.get(param, 'onCompleteWhenCan').exists)
			{
				var onCompleteWhenCan = module.exports._util.param.get(param, 'onCompleteWhenCan').value;

				delete param.onCompleteWhenCan;
				
				if (typeof onCompleteWhenCan == 'function')
				{
					onCompleteWhenCan(param);
				}
				else
				{
					module.exports._util.controller.invoke(onCompleteWhenCan, param)
				}
			}
		},

		param:
		{
			get: function(param, name, options)
			{
				if (param == undefined) {param = {}}
				if (options == undefined) {options = {}}
		
				var data = {exists: false, value: options.default};

				var split = options.split;
				var index = options.index;
				var remove = options.remove;	
				var set = options.set;
			
				if (param.hasOwnProperty(name))
				{
					if (param[name] != undefined) {data.value = param[name]};
					data.exists = true;

					if (index !== undefined && split === undefined) {split = '-'}

					if (split !== undefined)
					{
						if (param[name] !== undefined)
						{	
							data.values = param[name].split(split);

							if (index !== undefined)
							{
								if (index < data.values.length)
								{
									data.value = data.values[index];
								}
							}
						}	
					}

					if (remove) {delete param[name]};
					if (set) {param[name] = data.value};
				}

				return data;
			},

			set: function(param, key, value, options)
			{
				var onlyIfNoKey = false;

				if (module.exports._util.param.get(options, 'onlyIfNoKey').exists)
				{
					onlyIfNoKey = module.exports._util.param.get(options, 'onlyIfNoKey').value
				}

				if (param === undefined) {param = {}}

				if (param.hasOwnProperty(key))
				{
					if (!onlyIfNoKey) {param[key] = value};
				}
				else
				{
					param[key] = value;
				}
					
				return param
			}									
		},

		controller:
		{
			data: [],

			add: function (param)
			{
				if (_.isArray(param))
				{
					_.each(param, function(controller)
					{
						module.exports._util.controller._add(controller);
					});
				}
				else
				{
					module.exports._util.controller._add(param);
				}
			},

			_add: function (controller)
			{
				if (controller.name != undefined)
				{
					module.exports._util.controller.data.push(controller);

					module.exports.controller[controller.name] = controller.code;

					if (controller.alias != undefined)
					{
						if (module.exports[controller.alias] == undefined)
						{
							module.exports[controller.alias] = function(controllerParam, controllerData)
							{
								module.exports._util.controller.invoke(controller.name, controllerParam, controllerData)
							}
						}
					}
				}
			},

			invoke: function (param, controllerParam, controllerData)
			{
				var name;
				var returnData;

				if (_.isObject(param))
				{
					name = module.exports._util.param.get(param, 'name').value;
				}
				else
				{
					name = param;
				}
				
				if (name != undefined)
				{
					module.exports._util.testing.message('Invoking controller named: ' + name);

					module.exports._util.data.set(
					{
						controller: name,
						context: '_param',
						value: controllerParam
					});

					if (_.isFunction(module.exports.controller[name]))
					{
						module.exports._util.controller.data.last = name;
						returnData = module.exports.controller[name](controllerParam, controllerData);
					}
					else
					{
						returnData = 'No controller named ' + name;

						module.exports._util.testing.message('There is no controller named: ' + name);
						
						if (!_.isUndefined(controllerParam))
						{
							module.exports._util.testing.message(controllerParam);
						}

						if (!_.isUndefined(controllerData))
						{
							module.exports._util.testing.message(controllerData);
						}
					}
				}

				return returnData;
			},

			show: function (param)
			{
				var name;

				if (_.isObject(param))
				{
					name = module.exports._util.param.get(param, 'name').value;
				}
				else
				{
					name = param;
				}

				var message = [];
				
				if (name != undefined)
				{
					message.push('-');
					message.push(name + ':');
					if (module.exports.controller[name] != undefined)
					{
						message.push(module.exports.controller[name].toString());
					}
				}
				else
				{
					var controllers = _.sortBy(module.exports._util.controller.data, function (controller) {return controller.name});

					_.each(controllers, function (controller)
					{
						message.push('-');
						message.push(controller.name);
						if (controller.note != undefined)
						{
							message.push(controller.note);
						}
					});
				}

				module.exports._util.message(message);
			}
		},

		data:
		{
			reset: function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				
				if (controller != undefined)
				{
					module.exports.data[controller] = {}
				}
				
				if (scope != undefined)
				{
					module.exports.data[scope] = {}
				}
			},

			clear: function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				var context = module.exports._util.param.get(param, 'context').value;
				var name = module.exports._util.param.get(param, 'name').value;
				var value = module.exports._util.param.get(param, 'value').value;

				if (controller == undefined)
				{
					controller = scope;
				}

				if (controller != undefined)
				{
					if (context != undefined)
					{
						if (name != undefined)
						{
							if (module.exports.data[controller] != undefined)
							{
								if (module.exports.data[controller][context] != undefined)
								{
									delete module.exports.data[controller][context][name];
								}
							}
						}
						else 
						{
							if (module.exports.data[controller] != undefined)
							{
								delete module.exports.data[controller][context];
							}
						}	
					}
					else
					{
						if (name != undefined)
						{
							if (module.exports.data[controller] != undefined)
							{
								delete module.exports.data[controller][name];
							}
						}
						else
						{
							delete module.exports.data[controller];
						}
					}	
				}
			},

			set: function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				var context = module.exports._util.param.get(param, 'context').value;
				var name = module.exports._util.param.get(param, 'name').value;
				var value = module.exports._util.param.get(param, 'value').value;
				var merge = module.exports._util.param.get(param, 'merge', {default: false}).value;
				var data;

				if (controller == undefined)
				{
					controller = scope;
				}
				
				if (controller != undefined && controller != 'session')
				{		
					if (module.exports.data[controller] == undefined) {module.exports.data[controller] = {}}

					if (context != undefined)
					{
						if (module.exports.data[controller][context] == undefined) {module.exports.data[controller][context] = {}}
					}

					if (context != undefined)
					{
						if (name != undefined)
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller][context][name]))
							{
								module.exports.data[controller][context][name] = _.assign(module.exports.data[controller][context][name], value);
							}
							else
							{
								module.exports.data[controller][context][name] = value;
							}

							data = module.exports.data[controller][context][name];
						}
						else 
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller][context]))
							{
								module.exports.data[controller][context] = _.assign(module.exports.data[controller][context], value);
							}
							else
							{
								module.exports.data[controller][context] = value;
							}

							data = module.exports.data[controller][context];
						}	
					}
					else
					{
						if (name != undefined)
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller][name]))
							{
								module.exports.data[controller][name] = _.assign(module.exports.data[controller][name], value);
							}
							else
							{
								module.exports.data[controller][name] = value;
							}

							data = module.exports.data[controller][name]
						}
						else
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller]))
							{
								module.exports.data[controller] = _.assign(module.exports.data[controller], value);
							}
							else
							{
								module.exports.data[controller] = value;
							}

							data = module.exports.data[controller];
						}	
					}
				}
				
				return data
			},

			get: 	function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				var context = module.exports._util.param.get(param, 'context').value;
				var name = module.exports._util.param.get(param, 'name').value;
				var id = module.exports._util.param.get(param, 'id').value;
				var valueDefault = module.exports._util.param.get(param, 'valueDefault').value;
				var value;
				var clean = module.exports._util.param.get(param, 'clean', {"default": false}).value;
				var clone = module.exports._util.param.get(param, 'clone', {"default": false}).value;

				if (controller == undefined)
				{
					controller = scope;
				}

				if (controller != undefined)
				{
					if (module.exports.data[controller] != undefined)
					{
						if (context != undefined)
						{
							if (module.exports.data[controller][context] != undefined)
							{	
								if (name != undefined)
								{
									value = module.exports.data[controller][context][name];
								}
								else 
								{
									value = module.exports.data[controller][context];
								}
							}
						}
						else
						{
							if (name != undefined)
							{
								value = module.exports.data[controller][name];
							}
							else
							{
								value = module.exports.data[controller];
							}
						}
					}	
				}

				if (value != undefined && id != undefined)
				{
					value = _.find(value, function (v) {return v.id == id});
				}
				
				if (value == undefined && valueDefault != undefined)
				{
					value = valueDefault;
					param.value = value;
					module.exports._util.data.set(param);
				}

				if (_.isObject(value) && clean)
				{
					value = _.pickBy(value, function (value, key) {return !_.startsWith(key, '_')})
				}

				var valueReturn = (clone?_.clone(value):value);

				return valueReturn
			}
		},

		send: function (options, callBack)
		{
			//Send to other internet service

            var session = module.exports.data.session;
			var https = require('https');

			if (options == undefined) {options = {}}
					
			if (_.isUndefined(options.action)) {options.action = 'GET'};

			var headers = {};

			if (options.headers != undefined)
			{
				headers = _.assign(headers, options.headers)
			}

			var _requestData;

			if (_.isObject(options.data))
			{
				_requestData = JSON.stringify(options.data);

				headers['Content-Type'] = 'application/json';
				headers['Content-Length'] = Buffer.byteLength(_requestData);
			}

			options.hostname = options.hostname.replace('https://', '');
			options.hostname = options.hostname.replace('http://', '');

			if (!_.startsWith(options.path, '/'))
			{
				options.path = '/' + options.path;
			}

			var requestOptions =
			{
				hostname: options.hostname,
				port: 443,
				path: options.path,
				method: options.action,
				headers: headers
			};

			module.exports._util.testing.data(requestOptions, 'util-request-options');
			module.exports._util.testing.data(_requestData, 'util-request-data');

			var req = https.request(requestOptions, function(res)
			{
				res.setEncoding('utf8');

				var data = '';
				
				res.on('data', function(chunk)
				{
				  	data += chunk;
				});
				
				res.on('end', function ()
				{	
					module.exports._util.testing.data(data, 'util-request-response-data');

					dataResponse = JSON.parse(data);
			    	if (!_.isUndefined(callBack))
			    	{
			    		module.exports.invoke(callBack, options, {data: dataResponse})
			    	};
				});
			});

			req.on('error', function(error)
			{
				module.exports._util.testing.data(error, 'util-request-response-error');

			  	if (callBack)
			  	{
			  		module.exports.invoke(callBack, {error: error})
			  	};
			});

			req.end(_requestData)
		},

		clean: function (encodedString)
		{
			var translate_re = /&(nbsp|amp|quot|lt|gt);/g;
			var translate =
			{
				"nbsp":" ",
				"amp" : "&",
				"quot": "\"",
				"lt"  : "<",
				"gt"  : ">"
			};
			return encodedString.replace(translate_re, function(match, entity) {
			return translate[entity];
			}).replace(/&#(\d+);/gi, function(match, numStr) {
				var num = parseInt(numStr, 10);
				return String.fromCharCode(num);
			});
		},

        attachment:
        {
            upload: function (param, fileData)
            {
                var filename = module.exports._util.param.get(param, 'filename', {default: 'learn.txt'}).value;
                var object = module.exports._util.param.get(param, 'object').value;
                var objectContext = module.exports._util.param.get(param, 'objectContext').value;
                var base64 = module.exports._util.param.get(param, 'base64', {default: false}).value;
                var type = module.exports._util.param.get(param, 'type').value;
                var settings = module.exports.get({scope: '_settings'});
                var session = module.exports.data.session;

                if (base64)
                {
                    module.exports.cloud.invoke(
                    {
                        method: 'core_attachment_from_base64',
                        data:
                        {
                            base64: fileData.base64,
                            filename: filename,
                            object: object,
                            objectcontext: objectContext,
                        },
                        callback: module.exports._util.attachment.process,
                        callbackParam: param
                    });
                }
                else
                {
                    if (fileData.buffer != undefined)
                    {
                        var blob = Buffer.from(fileData.buffer)
                    }
                    else
                    {
                        var blob = fileData;
                    }
                    
                    var FormData = require('form-data');
                    var form = new FormData();
        
					console.log(filename)

                    form.append('file0', blob,
                    {
                        contentType: 'application/octet-stream',
                        filename: filename
                    });

					console.log('blob')
                    form.append('filename0', filename);
                    form.append('object', object);
                    form.append('objectcontext', objectContext);
                    form.append('method', 'ATTACH_FILE');
  
                    if (!_.isUndefined(type))
                    {
                        form.append('type0', type);
                    }

					var submitOptions =
					{
						port: 443,
						host: settings.entityos.hostname,
						path: '/rpc/attach/',
						headers: {'auth-sid': session.sid, 'auth-logonkey': session.logonkey}
					  }

                    form.submit(submitOptions, function(err, res)
                    {
                        res.resume();

                        res.setEncoding('utf8');
                        res.on('data', function (chunk)
                        {
                            var data = JSON.parse(chunk);
                            module.exports.attachment.process(param, data)
                        });
                    });
                }
            },

            process: function (param, response)
            {
                if (response.status == 'OK')
                {
                    var attachment;

                    if (_.has(response, 'data.rows'))
                    {
                        attachment = _.first(response.data.rows);
                    }
                    else
                    {
                        attachment = response;
                    }

                    var data =
                    {
                        attachment:
                        {
                            id: attachment.attachment,
                            link: attachment.attachmentlink,
                            href: '/download/' + attachment.attachmentlink
                        }
                    }
                }

                param = module.exports._util.param.set(param, 'data', data);

                module.exports._util.onComplete(param)
            }
        },

		generateRandomText: function (param)
		{
			var length = module.exports._util.param.get(param, 'length').value;
			var specialChars = module.exports._util.param.get(param, 'specialChars', {"default": false}).value;
			var charset = module.exports._util.param.get(param, 'charset').value;
			var referenceNumber = module.exports._util.param.get(param, 'referenceNumber', {"default": false}).value;

			var generatedText = '';

			if (referenceNumber)
			{
				charset = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
				if (_.isUndefined(length))
				{
					length = 6
				}

				for (let i = 0; i < length; i++) {
					const randomIndex = Math.floor(Math.random() * charset.length);
					generatedText += charset[randomIndex];
				}
			}
			else
			{
				if (_.isUndefined(length))
				{
					length = 16
				}

				if (_.isUndefined(charset))
				{
					charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

					if (specialChars)
					{
						charset += '!@#$%^&*()_+~`|}{[]:;?><,./-=';
					}
				}

				const crypto = require('crypto');

				const values = new Uint8Array(length);
				crypto.randomFillSync(values);
				
				for (let i = 0; i < length; i++) {
					generatedText += charset[values[i] % charset.length];
				}
			}

			return generatedText;
		},

		toBase58: function textToBase58(text)
		{
			const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

			const buffer = Buffer.from(text, 'utf8');
			
			let integer = BigInt(`0x${buffer.toString('hex')}`);
  
			let base58 = '';
			
			while (integer > 0n) {
				const remainder = integer % 58n;
				base58 = BASE58_ALPHABET[Number(remainder)] + base58;
				integer = integer / 58n;
			}

			for (let i = 0; i < buffer.length && buffer[i] === 0; i++) {
				base58 = '1' + base58;
			}
			
			return base58;
		}
    },

	add: function (param) {return module.exports._util.controller.add(param)},
	invoke: function (param, controllerParam, controllerData) {return module.exports._util.controller.invoke(param, controllerParam, controllerData)},
	set: function (param) {return module.exports._util.data.set(param)},
	get: function (param) {return module.exports._util.data.get(param)},
    upload: function (param, data) {return module.exports._util.attachment.upload(param, data)}
}
