// mydigitalstructure/mydigitalstructure-2.0.9.js
var _ = require('lodash')
var moment = require('moment');

module.exports = 
{
	VERSION: '2.0.9',

	data: {session: undefined},
	controller: {},

	init: function (callBack, settings)
	{
		if (settings == undefined)
		{	
			var fs = require('fs');

			fs.readFile('settings.json', function (err, buffer)
			{
				if (!err)
				{	
					var _settings = buffer.toString();
					settings = JSON.parse(_settings);
					module.exports.data._settings = settings;

					if (!_.isUndefined(settings))
					{
						if (!_.isUndefined(settings.mydigitalstructure))
						{
							settings = settings.mydigitalstructure;
						}

						module.exports.data.settings = settings;
					}

					module.exports.logon(callBack, settings)
				}
				else
				{
					console.log('ERROR! No settings.json file.')
				}
			});
		}
		else
		{
			module.exports.data._settings = settings;
					
			if (!_.isUndefined(settings.mydigitalstructure))
			{
				settings = settings.mydigitalstructure;
			}

			module.exports.data.settings = settings;
			module.exports.logon(callBack, settings);
		}	
	},

	logon: function (callBack, settings, param)
	{
		var https = require('https');
		var authenticationLevel = module.exports.data.settings.authenticationLevel;
		if (_.isUndefined(authenticationLevel)) {authenticationLevel = 1};
		var logon = settings.logon;
		var password = settings.password;
		var code = module.exports._util.param.get(param, 'code').value;

		var data = 
		{
			method: 'LOGON',
			logon: logon,
			localtime: moment().format('D MMM YYYY HH:mm:ss')
		}

		var requestData = 'logon=' + logon +
								'&localtime=' + moment().format('D MMM YYYY HH:mm:ss');

		if (authenticationLevel == 1)
		{
			requestData += '&passwordhash=' + module.exports._util.hash(logon + password);
		}
		else if (authenticationLevel == 2)
		{
			requestData += '&passwordhash=' + module.exports._util.hash(logon + password + module.exports.data.session.logonkey)
		}
		else if (authenticationLevel == 3)
		{
			requestData += '&passwordhash=' + module.exports._util.hash(logon + password + module.exports.data.session.logonkey + code)
		}

		module.exports._util.testing.data(requestData, 'mydigitalstructure.cloud.logon.request');
		
		var options =
		{
			hostname: settings.hostname,
			port: 443,
			path: '/rpc/logon/?method=LOGON',
			method: 'POST',
			headers:
			{
				'Content-Type': 'application/x-www-form-urlencoded',
				'Content-Length': requestData.length
			}
		};

		module.exports._util.testing.data(options, 'mydigitalstructure.cloud.send.request.options');

		var req = https.request(options, function(res)
		{
			res.setEncoding('utf8');

			var data = '';
			
			res.on('data', function(chunk)
			{
			  	data += chunk;
			});
			
			res.on('end', function ()
			{	
				module.exports._util.testing.data(data, 'mydigitalstructure.cloud.send.response.end');
				var _data = JSON.parse(data);
		    	module.exports.data.session = _data;
		    	if (_.isFunction(callBack)) {callBack({data: _data, settings: settings})};
			});
		});

		req.on('error', function(error)
		{
			module.exports._util.testing.data(error.message, 'mydigitalstructure.cloud.logon.response.error');
		 
		  	if (_.isFunction(callBack))
		  	{
		  		callBack({error: error})
		  	}
	  		else
	    	{
	    		module.exports._util.controller.invoke(callBack, options, error)
	    	}

		});

		module.exports._util.testing.data(requestData, 'mydigitalstructure.cloud.send.request.data');

		req.write(requestData);
		req.end()
	},

	logoff: function (param, response)
	{
		var uri = module.exports._util.param.get(param, 'uri', {"default": '/'}).value;
		var refresh = module.exports._util.param.get(param, 'refresh', {"default": true}).value;

		if (_.isUndefined(response))
		{
			module.exports.send(
			{
				url: '/rpc/core/?method=CORE_LOGOFF'
			},
			'',
			module.exports._util.logoff);
		}
		else
		{
			module.exports._util.testing.log('Logged off', 'mydigitalstructure.cloud._util.logoff');
			module.exports.data.session = undefined;
		}	
	},		

	send: function (options, data, callBack)
	{
		var https = require('https');
		var querystring = require('querystring');
		var settings = module.exports.data.settings;
		var session = module.exports.data.session;
		var requestData;

		if (_.isUndefined(data))
		{
			requestData = options.data
		}
		else
		{
			requestData = data
		}

		if (_.isObject(requestData))
		{
			data.sid = session.sid;
			data.logonkey = session.logonkey;

			requestData = JSON.stringify(requestData);
		}
		else
		{
			if (!_.isUndefined(requestData))
			{	
				requestData = requestData + '&sid=' + session.sid + '&logonkey=' + session.logonkey;
			}
			else
			{
				requestData = 'sid=' + session.sid + '&logonkey=' + session.logonkey;
			}
		}

		if (_.isUndefined(callBack))
		{
			callBack = options.callBack
		}					
				
		if (options.type == undefined) {options.type = 'POST'}
				
		var requestOptions =
		{
			hostname: settings.hostname,
			port: 443,
			path: options.url,
			method: options.type,
			headers:
			{
				'Content-Type': 'application/x-www-form-urlencoded',
				'Content-Length': requestData.length
			}
		};

		module.exports._util.testing.data(options, 'mydigitalstructure.cloud.send.request.options');

		var req = https.request(requestOptions, function(res)
		{
			res.setEncoding('utf8');

			var data = '';
			
			res.on('data', function(chunk)
			{
			  	data += chunk;
			});
			
			res.on('end', function ()
			{	
				module.exports._util.testing.data(data, 'mydigitalstructure.cloud.send.response.end');

				if (_.startsWith(data, '{') && settings.convert == 'true')
				{
					data = JSON.parse(data)
				}

		    	if (_.isFunction(callBack))
		    	{
		    		callBack(options, data)
		    	}
		    	else
		    	{
		    		module.exports._util.controller.invoke(callBack, options, data)
		    	}
			});
		});

		req.on('error', function(error)
		{
			module.exports._util.testing.message(error.message, 'mydigitalstructure.cloud.send.response.error');
		  	if (_.isFunction(callBack))
		  	{
		  		callBack(options, {error: error})
		  	}
	  		else
	    	{
	    		module.exports._util.controller.invoke(callBack, options, {error: error})
	    	}
		});

		module.exports._util.testing.data(requestData, 'mydigitalstructure.cloud.send.request.data');

		req.write(requestData);
		req.end()
	},

	cloud:
	{
		logon: function (param) {module.exports.logon(param)},
		logoff: function (param) {module.exports.logoff(param)},
		send: function (param, data, callBack) {module.exports.send(param, data, callBack)},
		invoke: function (param, data, callBack)
		{
			if (arguments.length == 1)
			{
				data = param.data;
				if (data == undefined) {data = param.fields}

				callBack = param.callback;
			}

			if (param.url == undefined && param.method != undefined)
			{
				param.url = '/rpc/' + param.method.split('_')[0] + '/?method=' + (param.method).toUpperCase();
			}

			if (param.url == undefined && param.object != undefined)
			{
				param.url = '/rpc/' + param.object.split('_')[0] + '/?method=' + (param.object).toUpperCase() + '_MANAGE';
			}

			if (_.isObject(data))
			{
				data = _.join(_.map(data, function (value, key) {return key + '=' + value}), '&')
			}

			module.exports.send(param, data, callBack)
		},
		create: function (param)
		{
			if (typeof arguments[0] != 'object')
			{
				param =
				{
					object: arguments[0],
					data: arguments[1],
					callback: arguments[2],
					mode: arguments[3]
				}
			}

			if (param.data == undefined && param.fields != undefined)
			{
				param.data = param.fields;
			}

			var endpoint = param.object.split('_')[0];

			if (_.isObject(param.data))
			{
				param.data = _.join(_.map(param.data, function (data, key) {return key + '=' + data}), '&')
			}

			module.exports.send(
			{
				type: 'post',
				url: '/rpc/' + endpoint + '/?method=' + (param.object).toUpperCase() + '_MANAGE'
			},
			param.data,
			param.callback);
		},
		update: function (param) {module.exports.cloud.create(param)},
		save: function (param) {module.exports.cloud.create(param)},

		delete: function (param)
		{
			if (typeof param.data != 'object')
			{
				param.data = {id: param.data}
			}

			param.endpoint = param.object.split('_')[0];
			param.data.remove = 1;

			module.exports.send(
			{
				type: 'post',
				url: '/rpc/' + endpoint + '/?method=' + (param.object).toUpperCase() + '_MANAGE'
			},
			param.data,
			param.callback);
		},

		retrieve: function (param)
		{
			param.endpoint = param.object.split('_')[0];	

			if (_.has(param.data, 'criteria'))
			{
				if (_.has(param.data.criteria, 'options'))
				{
					if (_.isUndefined(param.data.criteria.options.rows))
					{
						param.data.criteria.options.rows = mydigitalstructure._scope.app.options.rows;
					}
				}
				else
				{
					param.data.criteria.options = {rows: mydigitalstructure._scope.app.options.rows}
				}
			}
			else
			{
				if (param.data == undefined && (_.has(param, 'fields') || _.has(param, 'summaryFields')))
				{
					param.data =
					{ 
						criteria: 
						{
							"fields": [],
							"summaryFields": [],
							"filters": [],
							"sorts": [],
							"options": {},
							"customoptions": []
						}
					}

					if (_.has(param, 'fields')) { param.data.criteria.fields = param.fields }
					if (_.has(param, 'summaryFields')) { param.data.criteria.summaryFields = param.summaryFields }
					if (_.has(param, 'filters')) { param.data.criteria.filters = param.filters }
					if (_.has(param, 'sorts')) { param.data.criteria.sorts = param.sorts }
					if (_.has(param, 'options')) { param.data.criteria.options = param.options }
					if (_.has(param, 'customOptions')) { param.data.criteria.customOptions = param.customOptions }
				}
				else
				{
					param.managed = false;
				}
			}

			if (param.data.criteria.filters != undefined)
			{
				_.each(param.data.criteria.filters, function (filter)
				{
					if (filter.value1 == undefined && filter.value != undefined)
					{
						filter.value1 = filter.value
					}

					if (filter.name == undefined && filter.field != undefined)
					{
						filter.name = filter.field
					}
				})
			}

			param.type = 'post';
			param.url = '/rpc/' + param.endpoint + '/?method=' + (param.object).toUpperCase() + '_SEARCH';

			if (_.has(param.data, '_controller'))
			{
				param.data._controller = param.object + ':' + JSON.stringify(param.data.criteria.fields);
			}

			module.exports.send(
			{
				type: param.type,
				url: param.url
			},
			'criteria=' + JSON.stringify(param.data.criteria),
			param.callback);
		},

		search: function (param) {module.exports.cloud.retrieve(param)},
		query: function (param) {module.exports.cloud.retrieve(param)}
	},

	_util:
	{
		message: function (data, context, linebreak)
		{
			if (!_.isUndefined(context)) {console.log('[' + context + ']:')}

			if (_.isObject(data))
			{
				console.log(JSON.stringify(data, null, 4))
			}
			else
			{
				console.log(data);
			}

			if (!_.isUndefined(linebreak))
			{
				console.log(linebreak)
			}
		},

		testing:
		{
			message: function (message, context)
			{
				var settings = module.exports.data.settings.testing;

				if (_.isObject(settings))
				{
					if (settings.status == 'true')
					{
						if (!_.isUndefined(context)) {console.log('[' + context + ']:')}
						console.log(message);

						if (!_.isUndefined(settings.break))
						{
							console.log(settings.break)
						}
					}
				}	
			},

			data: function (data, context)
			{
				var settings = module.exports.data.settings.testing;

				if (_.isObject(settings))
				{
					if (settings.status == 'true' && settings.showData == 'true')
					{
						if (!_.isUndefined(context)) {console.log('[' + context + '][data]:')}

						if (_.isObject(data))
						{
							console.log(JSON.stringify(data, null, 4))
						}
						else
						{
							console.log(data);
						}

						if (!_.isUndefined(settings.break))
						{
							console.log(settings.break)
						}	
					}
				}	
			},

			status: function (status)
			{
				var settings = module.exports.data.settings.testing;

				if (_.isObject(settings))
				{
					if (_.isUndefined(status))
					{
						status = module.exports.data.settings.testing.status;
					}
					else
					{
						module.exports.data.settings.testing.status = status;
					}
				}

				return status
			}
		},

		search:  
		{	
			init: function ()
			{
				var criteria = 
				{
					"fields": [],
					"summaryFields": [],
					"filters": [],
					"sorts": [],
					"options": {},
					"customoptions": []
				}

				return criteria
			}
		},

		hash: function (data)
		{
			var crypto = require('crypto');
			return crypto.createHash('md5').update(data).digest("hex");
		},

		doCallBack: function()
		{
			var param, callback, data;

			if (typeof arguments[0] == 'function')
			{
				callback = arguments[0]
				param = arguments[1] || {};
				data = arguments[2]
			}
			else
			{
				param = arguments[0] || {};
				callback = param.callback;
				data = arguments[1];
				delete param.callback;
			}

			if (callback)
			{
				callback(param, data)
			};
		},

		onComplete: function (param)
		{
			if (module.exports._util.param.get(param, 'onComplete').exists)
			{
				var onComplete = module.exports._util.param.get(param, 'onComplete').value;

				if (module.exports._util.param.get(param, 'onCompleteWhenCan').exists)
				{
					param.onComplete = param.onCompleteWhenCan;
					delete param.onCompleteWhenCan;
				}	
				else
				{
					delete param.onComplete;
				}

				onComplete(param);
			}
			else if (module.exports._util.param.get(param, 'onCompleteWhenCan').exists)
			{
				var onCompleteWhenCan = module.exports._util.param.get(param, 'onCompleteWhenCan').value;

				delete param.onCompleteWhenCan;
			
				onCompleteWhenCan(param);
			}
		},

		param:
		{
			get: function(param, name, options)
			{
				if (param == undefined) {param = {}}
				if (options == undefined) {options = {}}
		
				var data = {exists: false, value: options.default};

				var split = options.split;
				var index = options.index;
				var remove = options.remove;	
				var set = options.set;
			
				if (param.hasOwnProperty(name))
				{
					if (param[name] != undefined) {data.value = param[name]};
					data.exists = true;

					if (index !== undefined && split === undefined) {split = '-'}

					if (split !== undefined)
					{
						if (param[name] !== undefined)
						{	
							data.values = param[name].split(split);

							if (index !== undefined)
							{
								if (index < data.values.length)
								{
									data.value = data.values[index];
								}
							}
						}	
					}

					if (remove) {delete param[name]};
					if (set) {param[name] = data.value};
				}

				return data;
			},

			set: function(param, key, value, options)
			{
				var onlyIfNoKey = false;

				if (module.exports._util.param.get(options, 'onlyIfNoKey').exists)
				{
					onlyIfNoKey = module.exports._util.param.get(options, 'onlyIfNoKey').value
				}

				if (param === undefined) {param = {}}

				if (param.hasOwnProperty(key))
				{
					if (!onlyIfNoKey) {param[key] = value};
				}
				else
				{
					param[key] = value;
				}
					
				return param
			}									
		},

		controller:
		{
			data: {note: {}},

			add: function (param)
			{
				if (_.isArray(param))
				{
					_.each(param, function(controller)
					{
						module.exports._util.controller._add(controller);
					});
				}
				else
				{
					module.exports._util.controller._add(param);
				}
			},

			_add: function (controller)
			{
				if (controller.name != undefined)
				{
					if (controller.note != undefined)
					{
						module.exports._util.controller.data.note[controller.name] = controller.note;
					}

					module.exports.controller[controller.name] = controller.code;

					if (controller.alias != undefined)
					{
						if (module.exports[controller.alias] == undefined)
						{
							module.exports[controller.alias] = function(controllerParam, controllerData)
							{
								module.exports._util.controller.invoke(controller.name, controllerParam, controllerData)
							}
						}
					}
				}
			},

			invoke: function (param, controllerParam, controllerData)
			{
				var name;
				var returnData;

				if (_.isObject(param))
				{
					name = module.exports._util.param.get(param, 'name').value;
				}
				else
				{
					name = param;
				}
				
				if (name != undefined)
				{
					module.exports._util.testing.message('Invoking controller named: ' + name);

					module.exports._util.data.set(
					{
						controller: name,
						context: '_param',
						value: controllerParam
					});

					if (_.isFunction( module.exports.controller[name]))
					{
						module.exports._util.controller.data.last = name;
						returnData =  module.exports.controller[name](controllerParam, controllerData);
					}
					else
					{
						returnData = 'No controller named ' + name;

						module.exports._util.testing.message('There is no controller named: ' + name);
						
						if (!_.isUndefined(controllerParam))
						{
							module.exports._util.testing.message(controllerParam);
						}

						if (!_.isUndefined(controllerData))
						{
							module.exports._util.testing.message(controllerData);
						}
					}
				}

				return returnData;
			}
		},

		data:
		{
			reset: function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				
				if (controller != undefined)
				{
					module.exports.data[controller] = {}
				}
				
				if (scope != undefined)
				{
					module.exports.data[scope] = {}
				}
			},

			clear: function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				var context = module.exports._util.param.get(param, 'context').value;
				var name = module.exports._util.param.get(param, 'name').value;
				var value = module.exports._util.param.get(param, 'value').value;

				if (controller == undefined)
				{
					controller = scope;
				}

				if (controller != undefined)
				{
					if (context != undefined)
					{
						if (name != undefined)
						{
							if (module.exports.data[controller] != undefined)
							{
								if (module.exports.data[controller][context] != undefined)
								{
									delete module.exports.data[controller][context][name];
								}
							}
						}
						else 
						{
							if (module.exports.data[controller] != undefined)
							{
								delete module.exports.data[controller][context];
							}
						}	
					}
					else
					{
						if (name != undefined)
						{
							if (module.exports.data[controller] != undefined)
							{
								delete module.exports.data[controller][name];
							}
						}
						else
						{
							delete module.exports.data[controller];
						}
					}	
				}
			},

			set: function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				var context = module.exports._util.param.get(param, 'context').value;
				var name = module.exports._util.param.get(param, 'name').value;
				var value = module.exports._util.param.get(param, 'value').value;
				var merge = module.exports._util.param.get(param, 'merge', {default: false}).value;
				var data;

				if (controller == undefined)
				{
					controller = scope;
				}
				
				if (controller != undefined && controller != 'session')
				{		
					if (module.exports.data[controller] == undefined) {module.exports.data[controller] = {}}

					if (context != undefined)
					{
						if (module.exports.data[controller][context] == undefined) {module.exports.data[controller][context] = {}}
					}

					if (context != undefined)
					{
						if (name != undefined)
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller][context][name]))
							{
								module.exports.data[controller][context][name] = _.assign(module.exports.data[controller][context][name], value);
							}
							else
							{
								module.exports.data[controller][context][name] = value;
							}

							data = module.exports.data[controller][context][name];
						}
						else 
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller][context]))
							{
								module.exports.data[controller][context] = _.assign(module.exports.data[controller][context], value);
							}
							else
							{
								module.exports.data[controller][context] = value;
							}

							data = module.exports.data[controller][context];
						}	
					}
					else
					{
						if (name != undefined)
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller][name]))
							{
								module.exports.data[controller][name] = _.assign(module.exports.data[controller][name], value);
							}
							else
							{
								module.exports.data[controller][name] = value;
							}

							data = module.exports.data[controller][name]
						}
						else
						{
							if (merge && _.isObject(value) && _.isObject(module.exports.data[controller]))
							{
								module.exports.data[controller] = _.assign(module.exports.data[controller], value);
							}
							else
							{
								module.exports.data[controller] = value;
							}

							data = module.exports.data[controller];
						}	
					}
				}
				
				return data
			},

			get: 	function (param)
			{
				var controller = module.exports._util.param.get(param, 'controller').value;
				var scope = module.exports._util.param.get(param, 'scope').value;
				var context = module.exports._util.param.get(param, 'context').value;
				var name = module.exports._util.param.get(param, 'name').value;
				var id = module.exports._util.param.get(param, 'id').value;
				var valueDefault = module.exports._util.param.get(param, 'valueDefault').value;
				var value;
				var clean = module.exports._util.param.get(param, 'clean', {"default": false}).value;
				var clone = module.exports._util.param.get(param, 'clone', {"default": false}).value;

				if (controller == undefined)
				{
					controller = scope;
				}

				if (controller != undefined)
				{
					if (module.exports.data[controller] != undefined)
					{
						if (context != undefined)
						{
							if (module.exports.data[controller][context] != undefined)
							{	
								if (name != undefined)
								{
									value = module.exports.data[controller][context][name];
								}
								else 
								{
									value = module.exports.data[controller][context];
								}
							}
						}
						else
						{
							if (name != undefined)
							{
								value = module.exports.data[controller][name];
							}
							else
							{
								value = module.exports.data[controller];
							}
						}
					}	
				}

				if (value != undefined && id != undefined)
				{
					value = _.find(value, function (v) {return v.id == id});
				}
				
				if (value == undefined && valueDefault != undefined)
				{
					value = valueDefault;
					param.value = value;
					module.exports._util.data.set(param);
				}

				if (_.isObject(value) && clean)
				{
					value = _.pickBy(value, function (value, key) {return !_.startsWith(key, '_')})
				}

				var valueReturn = (clone?_.clone(value):value);

				return valueReturn
			}
		}
	},

	add: function (param) {module.exports._util.controller.add(param)},
	invoke: function (param, controllerParam, controllerData) {module.exports._util.controller.invoke(param, controllerParam, controllerData)},
	set: function (param) {return module.exports._util.data.set(param)},
	get: function (param) {return module.exports._util.data.get(param)}
}
